File: src/main/java/StackBasedTowerOfHanoi.java
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.EnumSet;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class StackBasedTowerOfHanoi {

  Map<StackType, Deque<Integer>> stacks;

  public StackBasedTowerOfHanoi(Map<StackType, Deque<Integer>> stacks) {
    this.stacks = stacks;
  }

  private static Map<StackType, Deque<Integer>> initStacks(int size) {
    var stacks = createStacks(size);
    var sourceStack = stacks.get(StackType.SRC);

    IntStream.rangeClosed(1, size).forEach(sourceStack::addLast);

    return stacks;
  }

  private static Map<StackType, Deque<Integer>> createStacks(int maxSize) {
    return EnumSet.allOf(StackType.class).stream().collect(Collectors.toMap(Function.identity(), t -> new ArrayDeque<>(maxSize)));
  }

  public static void main(String[] args) {
    var size = 4;
    var towerOfHanoi = new StackBasedTowerOfHanoi(StackBasedTowerOfHanoi.initStacks(size));

    System.out.println("Start: " + towerOfHanoi.stacks);
    towerOfHanoi.solve(size);
    System.out.println("End: " + towerOfHanoi.stacks);
  }

  public void solve(int num) {
    solveInternal(num, StackType.SRC, StackType.DEST, StackType.AUX);
  }

  private void solveInternal(int num, StackType source, StackType destination, StackType auxiliary) {

    if (num > 0) {
      solveInternal(num - 1, source, auxiliary, destination);

      var value = this.stacks.get(source).pop();
      this.stacks.get(destination).push(value);
      System.out.println(String.format("Move disk %d from %s to %s -> %s", value, source, destination, this.stacks));

      solveInternal(num - 1, auxiliary, destination, source);
    }
  }

  enum StackType {
    SRC,
    DEST,
    AUX
  }
}

File: src/main/kotlin/StackBasedTowerOfHanoi.kt
fun <T> ArrayDeque<T>.push(element: T) = addFirst(element)
fun <T> ArrayDeque<T>.pop() = removeFirstOrNull()

class StackBasedTowerOfHanoi(private val stacks: Map<StackType, ArrayDeque<Int>>) {
    fun solve(num: Int) {
        solveInternal(num, StackType.SRC, StackType.DEST, StackType.AUX)
    }

    private fun solveInternal(
        num: Int,
        source: StackType,
        destination: StackType,
        auxiliary: StackType
    ) {
        if (num > 0) {
            solveInternal(num - 1, source, auxiliary, destination)

            val value = stacks[source]?.pop() as Int
            stacks[destination]?.push(value)

            println("Move disk $value from $source to $destination -> $stacks")

            solveInternal(num - 1, auxiliary, destination, source)
        }
    }

    enum class StackType {
        SRC, DEST, AUX
    }

    companion object {
        private fun initStacks(size: Int): Map<StackType, ArrayDeque<Int>> {
            val stacks = createStacks(size)

            val sourceStack = stacks[StackType.SRC]
            for (i in size downTo 1) sourceStack?.push(i)

            return stacks
        }

        private fun createStacks(maxSize: Int): Map<StackType, ArrayDeque<Int>> {
            return StackType.values().associateWith { ArrayDeque(maxSize) }
        }

        @JvmStatic
        fun main(args: Array<String>) {
            val size = 4
            val towerOfHanoi = StackBasedTowerOfHanoi(initStacks(size))
            println("Start: " + towerOfHanoi.stacks)
            towerOfHanoi.solve(size)
            println("End: " + towerOfHanoi.stacks)
        }
    }
}
